Packages/com.unity.inputsystem/InputSystem/Editor/AssetImporter/InputActionImporter.cs

```diff
@@ -256,6 +256,39 @@ public static void CreateInputAsset()
             ProjectWindowUtil.CreateAssetWithContent("New Controls." + InputActionAsset.Extension,
                 InputActionAsset.kDefaultAssetLayoutJson, InputActionAssetIconLoader.LoadAssetIcon());
         }
+
+        // When an action asset is renamed, copied, or moved in the Editor, the "Name" field in the JSON will
+        // hold the old name and won't match what's in memory (asset looks "dirty"). To work around this, we
+        // must flush the updated JSON to the file whenever this operation occurs.
+        // https://jira.unity3d.com/browse/ISXB-749
+        private class PostProcessAssetModifications : AssetPostprocessor
```

Comment:
Interesting this was the missing piece for this particular bug. I stumbled upon this myself while working on https://jira.unity3d.com/browse/ISX-1865 but from another angle. The missing functionality I identified is what is being done here for the "legacy" IMGUI based editor: https://github.com/Unity-Technologies/InputSystem/blob/develop/Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionEditorWindow.cs#L856-L914
Also of relevance is: https://github.com/Unity-Technologies/InputSystem/blob/develop/Packages/com.unity.inputsystem/InputSystem/Editor/AssetEditor/InputActionAssetManager.cs
I haven't fully landed in how this should work given that the functionality present there seems mainly focused on updating the editor if the file on disc changes while the editor content is unmodified. 
So in addition to the scenarios you have covered here reflecting asset modifications via files it is relevant to consider what to expect for scenarios where the file modifications happen and there is also in-memory modifications of the asset in Unity editors. I am afraid that is not possible to solve in a good way apart from selecting either update-from-disc or overwrite-to-disc. Due to this I only comment for now and try to think about this a bit more and get back to you tomorrow. Feel free to share your perspective as well based on existing implementation above. It clearly looks like the UITK editor was constructed overlooking these aspects from the older editor implementation.

Reply:
Unfortunately I've been unable to find a real solution to the Warnings logged to the console (which Paulinus pointed out) when creating a new asset and in a couple other scenarios. This leads me to believe this current approach is wrong, and we should not be modifying the asset file during the import process. So, even if we decide the fix itself is correct, it needs to be executed somewhere else and not during import.

I do see this more as a fundamental problem with InputActionAsset handling and not just an oversight with UITK's implementation. I did try a scenario in which the asset file was moved while the ActionEditor window was open (with unsaved changes), resulting in a FileNotFound exception with the in-memory changes failing to save.

However, I don't think this is the most important problem to solve right now, and as you said there isn't a good solution. It might be better to implement a work-around fix for the UITK ActionsEditor and revisit the larger issue later.